apps/withdraw-balancer-v2/src/utils/readPairsData.ts

```diff
@@ -0,0 +1,117 @@
+import { BigNumber } from "ethers";
+import type { Address, PublicClient } from "viem";
+import { uniswapV2PairAbi } from "./abis/uniswapV2PairAbi";
+
+interface PairData {
+  address: string;
+  userBalance: BigNumber;
+  price0CumulativeLast: BigNumber;
+  price1CumulativeLast: BigNumber;
+  reserve0: BigNumber;
+  reserve1: BigNumber;
+  totalSupply: BigNumber;
+}
+
+export async function readPairsData(
+  userAddress: string,
+  pairAddresses: string[],
+  client: PublicClient,
+): Promise<PairData[]> {
+  try {
```

Comment:
Make sure you consider balancesDiff from the frame context when you use this user balance.